lib/pbench/server/api/resources/server_configuration.py

```diff
@@ -0,0 +1,152 @@
+        super().__init__(
+            config,
+            logger,
+            ApiSchema(
+                API_METHOD.PUT,
+                API_OPERATION.UPDATE,
+                uri_schema=Schema(
+                    Parameter("key", ParamType.KEYWORD, keywords=ServerConfig.KEYS)
+                ),
+                # TODO: We aspire to be flexible about how a config option is
+                # specified on a PUT. While the "config" parameter is specifically
+                # intended to support a full JSON body that sets or updates all
+                # options at once, for individual options using the "value"
+                # query parameter is probably easier... but the "value" body
+                # parameter also allows specifying the value in a body.
+                #
+                #   PUT /server/configuration/dataset-lifetime?value=2y
+                #   PUT /server/configuration/dataset-lifetime
+                #       {"value": "2y"}
+                #   PUT /server/configuration/dataset-lifetime ???
+                #       {"config": {"dataset-lifetime": "2y"}}
+                #
+                # Some options may not make sense: need to think about this.
+                query_schema=Schema(Parameter("value", ParamType.STRING)),
+                body_schema=Schema(
+                    Parameter("config", ParamType.JSON, keywords=ServerConfig.KEYS),
+                    Parameter("value", ParamType.JSON),
+                ),
+                authorization=API_AUTHORIZATION.ADMIN,
+            ),
+            ApiSchema(
+                API_METHOD.GET,
+                API_OPERATION.READ,
+                uri_schema=Schema(
+                    Parameter("key", ParamType.KEYWORD, keywords=ServerConfig.KEYS)
+                ),
+                authorization=API_AUTHORIZATION.NONE,
+            ),
+            always_enabled=True,
+        )
+
+    def _get(self, params: ApiParams, request: Request) -> Response:
+        """
+        Get the values of server configuration parameters.
+
+        Args:
+            params: API parameters
+            request: The original Request object containing query parameters
+
+        GET /api/v1/server/configuration?key=config1&key=config2
+        """
+
+        key = params.uri.get("key")
+        try:
+            if key:
+                s = ServerConfig.get(key)
+                return jsonify({key: s.value if s else None})
+            else:
+                return jsonify(ServerConfig.get_all())
+        except ServerConfigError as e:
+            raise APIAbort(HTTPStatus.INTERNAL_SERVER_ERROR, str(e))
+
+    def _put(self, params: ApiParams, _) -> Response:
+        """
+        Set or modify the values of server configuration keys.
+
+        PUT /api/v1/server/configuration
+        {
+            "config": {
+                "dataset-lifetime": 10,
+                "server-state": "running"
+            }
+        }
```

Comment:
If the request is `PUT /api/v1/server/configuration` (no trailing slash or key value), then what does `params.uri.get("key")` return?  If it's false-y, then I think it is reasonable for the code to expect the request body to contain a mapping of server keys to values, and we don't need the `"config"` intermediary.  If it returns a key value, then we can expect that the body contains an object with a `"value"` key if that's how we want to receive the value.

From where I sit, `PUT /api/v1/server/configuration` should be a completely different method from `PUT /api/v1/server/configuration/<key>`.  Yes, they are semantically related, but they are different methods.  The fact that Flask routes them to the same code is not a problem if that code can tell which one was requested.  And, so the two "methods" can have different behaviors.  And, we don't need to encumber them with extra levels in the JSON body just because they end up being handled by the same code.